crates/wasm-smith/src/core.rs

```diff
@@ -1001,69 +1019,90 @@ impl Module {
 
     fn arbitrary_composite_type(&mut self, u: &mut Unstructured) -> Result<CompositeType> {
         use CompositeInnerType as CT;
-        let shared = false; // TODO: handle shared
+        let shared = self.arbitrary_shared(u)?;
+
         if !self.config.gc_enabled {
             return Ok(CompositeType {
                 shared,
-                inner: CT::Func(self.arbitrary_func_type(u)?),
+                inner: CT::Func(self.arbitrary_func_type(u, shared)?),
             });
         }
 
         match u.int_in_range(0..=2)? {
             0 => Ok(CompositeType {
                 shared,
-                inner: CT::Array(ArrayType(self.arbitrary_field_type(u)?)),
+                inner: CT::Array(ArrayType(self.arbitrary_field_type(u, shared)?)),
             }),
             1 => Ok(CompositeType {
                 shared,
-                inner: CT::Func(self.arbitrary_func_type(u)?),
+                inner: CT::Func(self.arbitrary_func_type(u, shared)?),
             }),
             2 => Ok(CompositeType {
                 shared,
-                inner: CT::Struct(self.arbitrary_struct_type(u)?),
+                inner: CT::Struct(self.arbitrary_struct_type(u, shared)?),
             }),
             _ => unreachable!(),
         }
     }
 
-    fn arbitrary_struct_type(&mut self, u: &mut Unstructured) -> Result<StructType> {
+    fn arbitrary_struct_type(
+        &mut self,
+        u: &mut Unstructured,
+        must_share: bool,
+    ) -> Result<StructType> {
         let len = u.int_in_range(0..=20)?;
         let mut fields = Vec::with_capacity(len);
         for _ in 0..len {
-            fields.push(self.arbitrary_field_type(u)?);
+            fields.push(self.arbitrary_field_type(u, must_share)?);
         }
         Ok(StructType {
             fields: fields.into_boxed_slice(),
         })
     }
 
-    fn arbitrary_field_type(&mut self, u: &mut Unstructured) -> Result<FieldType> {
+    fn arbitrary_field_type(
+        &mut self,
+        u: &mut Unstructured,
+        must_share: bool,
+    ) -> Result<FieldType> {
         Ok(FieldType {
-            element_type: self.arbitrary_storage_type(u)?,
+            element_type: self.arbitrary_storage_type(u, must_share)?,
             mutable: u.arbitrary()?,
         })
     }
 
-    fn arbitrary_storage_type(&mut self, u: &mut Unstructured) -> Result<StorageType> {
+    fn arbitrary_storage_type(
+        &mut self,
+        u: &mut Unstructured,
+        must_share: bool,
```

Comment:
One thing we've done in the past historically is to avoid threading through parameters like this and storing them on `Self`. Would that perhaps make sense for this `must_share` field? That'd help many callers I think since ideally the shared-ness would, in a perfect world, integrate in a few locations but wouldn't permeate throughout.

Reply:
I don't think we can: if we set it on `Self` it would be a module-level decision but we actually need a type-by-type decision. It boils down to: "we just generated a `shared` object and we need to ensure that any types inside it are also `shared`" &mdash; thus, `must_share = true`. The unshared case is simpler but we also want to generate it: "we just generated an unshared object and it doesn't matter if the types inside are shared or not" &mdash; `must_share = false`, which allows us to choose later.

There is another way to do this: we could "bubble up" the shared-ness from down below instead of pushing it down from above. I asked @fitzgen what he preferred and he picked the "push down" approach... so here we are.

Reply:
And to clarify: if it's a module-level flag we would have to set it before calling down and then unset it on the way back up. I figured threading it is more clear.

Reply:
Ah yes sorry I meant for this to be a you-set-it-then-unset-it kind of field where it's not global-per-module, just stored on the side to avoid threading everywhere. This pattern is used for fields like `max_type_limit` I believe so I think this'll fit reasonably well into that pattern. 

I agree that passing arguments is more explicit, but it's a fair bit of noise much of the time since it's just being forwarded around. The integration with `arbitrary_shared` I think might also make the most sense by putting it into `Module` itself.

Reply:
Take a look at f8b9edb. After that change I ran it for more than an hour without failure racking up ~14m iterations. (Hope we're still emitting shared stuff!).